packages/@aws-cdk/cdk/lib/cloudformation/arn.ts

```diff
@@ -44,22 +45,51 @@ export class Arn extends CloudFormationToken {
             values.push(components.resourceName);
         }
 
-        return new Arn(new FnConcat(...values));
+        return new FnConcat(...values).toString();
     }
 
     /**
      * Given an ARN, parses it and returns components.
      *
-     * The ARN it will be parsed and validated. The separator (`sep`) will be
-     * set to '/' if the 6th component includes a '/', in which case, `resource`
-     * will be set to the value before the '/' and `resourceName` will be the
-     * rest. In case there is no '/', `resource` will be set to the 6th
-     * components and `resourceName` will be set to the rest of the string.
+     * If the ARN is a concrete string, it will be parsed and validated. The
+     * separator (`sep`) will be set to '/' if the 6th component includes a '/',
+     * in which case, `resource` will be set to the value before the '/' and
+     * `resourceName` will be the rest. In case there is no '/', `resource` will
+     * be set to the 6th components and `resourceName` will be set to the rest
+     * of the string.
+     *
+     * If the ARN includes tokens (or is a token), the ARN cannot be validated,
+     * since we don't have the actual value yet at the time of this function
+     * call. You will have to know the separator and the type of ARN. The
+     * resulting `ArnComponents` object will contain tokens for the
+     * subexpressions of the ARN, not string literals. In this case this
+     * function cannot properly parse the complete final resourceName (path) out
+     * of ARNs that use '/' to both separate the 'resource' from the
+     * 'resourceName' AND to subdivide the resourceName further. For example, in
+     * S3 ARNs:
+     *
+     *      arn:aws:s3:::my_corporate_bucket/path/to/exampleobject.png
+     *
+     * After parsing the resourceName will not contain
+     * 'path/to/exampleobject.png' but simply 'path'. This is a limitation
+     * because there is no slicing functionality in CloudFormation templates.
+     *
+     * @param sep The separator used to separate resource from resourceName
+     * @param hasName Whether there is a name component in the ARN at all. For
+     * example, SNS Topics ARNs have the 'resource' component contain the topic
+     * name, and no 'resourceName' component.
+     *
+     * @returns an ArnComponents object which allows access to the various
+     * components of the ARN.
      *
      * @returns an ArnComponents object which allows access to the various
      *          components of the ARN.
      */
-    public static parse(arn: string): ArnComponents {
+    public static parse(arn: string, sepIfToken: string = '/', hasName: boolean = true): ArnComponents {
+        if (isToken(arn)) {
+            return this.parseToken(new CloudFormationToken(arn), sepIfToken, hasName);
```

Comment:
Why are you allowed to type `this` here? Shouldn't this be `ArnUtils.parseToken()` ?

Reply:
Fixed. I think it works because both are static, but still..